block-layer/graphsync/graphsync.md

```diff
@@ -59,6 +59,99 @@ type GraphSyncNet interface {
 
 ## Network Messages
 
+### Protocol Version 1.1.0
+
+Graphsync network messages are encoded in DAG-CBOR. They have the following schema
+
+```ipldsch
+type GraphSyncExtensions {string:Any}
+type GraphSyncRequestID int
+type GraphSyncPriority int
+
+type GraphSyncMetadatum struct {
+  Link         &Any
+	BlockPresent bool
+}
+type GraphSyncMetadata [GraphsyncMetadatum]
+
+type GraphSyncResponseCode enum {
+	# RequestAcknowledged means the request was received and is being worked on.
+	| RequestAcknowledged ("10")
+	# AdditionalPeers means additional peers were found that may be able
+	# to satisfy the request and contained in the extra block of the response.
+	| AdditionalPeers ("11")
+	# NotEnoughGas means fulfilling this request requires payment.
+	| NotEnoughGas ("12")
+	# OtherProtocol means a different type of response than GraphSync is
+	# contained in extra.
+	| OtherProtocol ("13")
+	# PartialResponse may include blocks and metadata about the in progress response
+	# in extra.
+	| PartialResponse ("14")
+	# RequestPaused indicates a request is paused and will not send any more data
+	# until unpaused
+	| RequestPaused ("15")
+
+	# Success Response Codes (request terminated)
+
+	# RequestCompletedFull means the entire fulfillment of the GraphSync request
+	# was sent back.
+	| RequestCompletedFull ("20")
+	# RequestCompletedPartial means the response is completed, and part of the
+	# GraphSync request was sent back, but not the complete request.
+	| RequestCompletedPartial ("21")
+
+	# Error Response Codes (request terminated)
+
+	# RequestRejected means the node did not accept the incoming request.
+	| RequestRejected ("30")
+	# RequestFailedBusy means the node is too busy, try again later. Backoff may
+	# be contained in extra.
+	| RequestFailedBusy ("31")
+	# RequestFailedUnknown means the request failed for an unspecified reason. May
+	# contain data about why in extra.
+	| RequestFailedUnknown ("32")
+	# RequestFailedLegal means the request failed for legal reasons.
+	| RequestFailedLegal ("33")
+	# RequestFailedContentNotFound means the respondent does not have the content.
+	| RequestFailedContentNotFound ("34")
+	# RequestCancelled means the responder was processing the request but decided to top, for whatever reason
+	| RequestCancelled ("35")
+} representation int
+
+type GraphSyncRequest struct {
+  Id          GraphSyncRequestID                 # unique id set on the requester side
+  Root        &Any                # a CID for the root node in the query
```

Comment:
This is kinda an interesting one.  It's a link, alright, so this line as stated is certainly valid.  But I'd like to think out loud about it for a moment.

While this is a link, it's not a link we'd ever really mean to traverse when processing the graphsync messages.  It's something we expect the agent processing this message to pull out -- _as_ a link, _not_ immediately traversing it -- and then do some logic before it dives in to operations that load that link.

It was just "bytes" in the old protobuf, and that worked fine because of the above.  And it seems we could readily still have it be `Bytes` in this revamp, too.  In the old implementation with protobuf, the application logic did a step to reify those bytes into a CID.  We could have that step be explicit in a new implementation too, if we wanted to.

This doesn't seem to make a huge difference either way.  I guess the main discernible difference is that if someone ran a selector with a bunch of wildcards in it _over the GraphSyncRequest itself_, if this field is `&Any`, they might get a mouthful back, and if it's `Bytes`, the results wouldn't recurse as far.  But does this matter?  Not really, as far as I can imagine.  (I don't know why someone would do that; and even if they did, selectors are full of mechanisms that can be used to control this.)

Alright, that's it for thinking out loud.  This seems fine.

Reply:
hah, that is an interesting point! probably doesn't matter here but worth keeping in mind that viewing this schema from a certain lens might lead to a misunderstanding about intent. `&Any` does provide some constraints on _what_ the bytes can be, mainly the leading varints must make for a valid CID, but beyond that if you're not traversing then it really is bytes.